src/backends/perf_pt/mod.rs

```diff
@@ -179,39 +197,40 @@ impl PerfPTTracer {
     }
 }
 
+impl Drop for PerfPTTracer {
+    fn drop(&mut self) {
+        let res = unsafe { perf_pt_free_tracer(self.tracer_ctx) };
```

Comment:
This worries me a bit. It looks to me like `perf_pt_free_tracer` is the sort of call which will fail at some point... and then... boom, the whole program dies. Can we / should we make users manually free things?

Reply:
We could, but it's not very Rusty, is it?

Reply:
[FWIW, I don't expect this to ever fail]

Reply:
I don't agree: I think it *will* fail at some point. Every Unix API that can lead to failure does, I've learnt, lead to failure at some point in practise. I know that `Drop` is easy to use, but I think we'll end up having some sort of manual way of doing things anyway in the long run, so we might as well bite the bullet now.

Reply:
Why would it fail? If it did, it would certainly be a bug on our part.

This will only fail if `munmap`, `close`, `free`, or `pthread_join` fails.

If one of those were to fail, what would the user do with the error, which basically says "something really weird happened"?

Reply:
`EINTR` if nothing else...

Reply:
It's true that close can return `EINTR` and it's never been clear what you should do if it does:
https://lwn.net/Articles/576478/

We can ask the user to explicitly check a result from a manual free, but at which point all bets are kind of off.

So what do you think we should we do? Still make it explicit?

Reply:
There are other ways for things to go wrong, but this is one. I've seen `close` fail too. So, yes, I really fear we have to make this explicit :/

Reply:
I meant `pthread_join` -- sorry!

Reply:
OK, let's do that. Although I think it's a shame that we couldn't make it more Rusty.

Reply:
This is the same sort of problem ones have with GC: finalisers look really useful, until you realise that at times like this, they're not quite the right thing.